File: tests/test_foma_fst/test_foma_fst.cpp
#include <climits>
#include <cstdlib>
#include <cstring>
#include <ctime>
#include <fstream>
#include <string>

#include "gtest/gtest.h"
#include "pyis/ops/text/foma_fst.h"
TEST(TestFomaFst, Basic) {
    pyis::ops::FomaFst fst("tests/test_foma_fst/data/recognize_latin.fst");

    ASSERT_EQ(fst.ApplyDown("avoid writing ad hoc code"), "avoid writing <latin>ad hoc</latin> code");
}

TEST(TestFomaFst, Serilize) {
    system("mkdir tmp");
    pyis::ops::FomaFst fst("tests/test_foma_fst/data/recognize_latin.fst");
    ASSERT_EQ(fst.ApplyDown("avoid writing ad hoc code"), "avoid writing <latin>ad hoc</latin> code");
    std::ofstream fout("tmp/FomaFstSerilize.fst", std::ios::binary);
    fst.SaveBinary(fout);
    fout.close();
    std::ifstream fin = std::ifstream("tmp/FomaFstSerilize.fst", std::ios::binary);
    fst.LoadBinary(fin);
    fin.close();
    ASSERT_EQ(fst.ApplyDown("avoid writing ad hoc code"), "avoid writing <latin>ad hoc</latin> code");
}

TEST(TestFomaFst, CompileFromStr) {
    system("mkdir tmp");
    pyis::ops::FomaFst::CompileFromStr(R""""(regex "ad hoc" -> %<latin%> ... %<%/latin%> ;
apply down avoid writing ad hoc code
save stack tmp/FomaFstCompileFromStr.fst )"""");
    pyis::ops::FomaFst fst("tmp/FomaFstCompileFromStr.fst");
    ASSERT_EQ(fst.ApplyDown("avoid writing ad hoc code"), "avoid writing <latin>ad hoc</latin> code");
    // run second time to test memory leak
    pyis::ops::FomaFst::CompileFromStr(R""""(regex "ad hoc" -> %<latin%> ... %<%/latin%> ;
apply down avoid writing ad hoc code
save stack tmp/FomaFstCompileFromStr.fst )"""");
}

File: py/pyis-python/tests/test_foma_fst/test_foma_fst.py
import unittest
from pyis.python import ops
import os
import sys

class TestFomaFst(unittest.TestCase):
    def test_basic(self):
        fst = ops.FomaFst(os.path.join(os.path.dirname(__file__), 'recognize_latin.fst'))
        query = "avoid writing ad hoc code"
        expected_res = "avoid writing <latin>ad hoc</latin> code"
        self.assertEqual(expected_res, fst.apply_down(query))

    def test_compile(self):
        os.makedirs("./tmp/", exist_ok = True)
        save_path = str(Path('./tmp/recognize_latin_saved.fst'))
        script_str = r"""
regex "ad hoc" -> %<latin%> ... %<%/latin%> ;
apply down avoid writing ad hoc code
save stack $path
"""
        script_str = script_str.replace("$path", save_path)
        print(script_str)
        ops.FomaFst.compile_from_str(script_str)
        fst = ops.FomaFst(save_path)
        query = "avoid writing ad hoc code"
        expected_res = "avoid writing <latin>ad hoc</latin> code"
        self.assertEqual(expected_res, fst.apply_down(query))

if __name__ == "__main__":
    unittest.main()
